src/core/async-work-cache.ts

```diff
@@ -50,19 +64,21 @@ export class AsyncWorkCache<K, V> {
     return promise;
   }
 
-  delete (key: K) {
-    this._keyToResultMap.delete(key);
-  }
-
-  clear() {
-    this._keyToResultMap.clear();
-  }
-
-  set(key: K, value: V) {
-    this._keyToResultMap.set(key, Promise.resolve(value));
+  /**
+   * Returns a copy of the underlying map which can be mutated as necessary
+   * and then used to create another cache.
+   */
+  fork(): Map<K, Promise<V>> {
```

Comment:
Why don't you make `AsyncWorkCache` implement `[Symbol.iterator]` so that it can be an argument to `new Map()`:

```ts
[Symbol.iterator]() {
  return this._keyToResultMap.entries();
}
```

Then above

```ts
const parsedDocumentPromises = new Map(filterEntries(this.parsedDocumentPromises, documentPaths));
// etc

function* filterEntries(entries: Iterable<K, V>, filteredKeys: Set<K>) {
  for (const entry of entries) {
    if (!filteredKeys.has(entry[0])) {
      yield entry;
    }
  }
}
```


I think that might make the forking steps a little clearer. Not a huge deal though.